client/keystore/src/lib.rs

```diff
@@ -275,29 +300,23 @@ impl Store {
 
 		Ok(public_keys)
 	}
-
-	/// Get public keys of all stored keys that match the key type.
-	///
-	/// This will just use the type of the public key (a list of which to be returned) in order
-	/// to determine the key type. Unless you use a specialized application-type public key, then
-	/// this only give you keys registered under generic cryptography, and will not return keys
-	/// registered under the application type.
-	pub fn public_keys<Public: AppPublic>(&self) -> Result<Vec<Public>> {
-		self.public_keys_by_type::<Public::Generic>(Public::ID)
-			.map(|v| v.into_iter().map(Into::into).collect())
-	}
-
-	/// Returns the file path for the given public key and key type.
-	fn key_file_path(&self, public: &[u8], key_type: KeyTypeId) -> Option<PathBuf> {
-		let mut buf = self.path.as_ref()?.clone();
-		let key_type = hex::encode(key_type.0);
-		let key = hex::encode(public);
-		buf.push(key_type + key.as_str());
-		Some(buf)
-	}
 }
 
 impl BareCryptoStore for Store {
+	fn keys(
+		&self,
+		id: KeyTypeId
+	) -> std::result::Result<Vec<CryptoTypePublicPair>, TraitError> {
+		let raw_keys = self.raw_public_keys(id)?;
+		Ok(raw_keys.into_iter()
+		   .map(|k| vec![
+			   CryptoTypePublicPair(sr25519::CRYPTO_ID, k.clone()),
+			   CryptoTypePublicPair(ed25519::CRYPTO_ID, k.clone())
+		   ])
+		   .flatten()
+		   .collect())
```

Comment:
I feel like [`fold`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold) would be a bit nicer here, rather than allocating a bunch of 2-elem vectors for each items, just do
```suggestion
		   .fold(Vec::new(), |mut v, k| {
		   		v.push(CryptoTypePublicPair(sr25519::CRYPTO_ID, k.clone()));
		   		v.push(CryptoTypePublicPair(ed25519::CRYPTO_ID, k.clone()));
		   		v
		   	});
```